make functionalization work better with subclasses

bdhirsh opened the pull request:
This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```


Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73442
* __->__ #73441

Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

bdhirsh committed:
make functionalization work better with subclasses

[ghstack-poisoned]

ezyang commented:
> One potential downside is that we would no longer be able to error on code like this:

I guess this is fine, if users who require functional only kernels appropriately error if they get an inplace kernel. Then it isn't functionalization's responsibility to raise the error, it's downstream, which is good because it means that inplaces can be resolved at different levels in the reentrant functionalization.

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -40,15 +40,22 @@ def is_tensor_like(a: Union[Argument, TensorOptionsArguments, SelfArgument]) ->
 # (1) a string containing all of the logic that does the unwrapping
 # (2) a context, to be used by translate(), with all of the relevant bindings.
-def unwrap_tensor_args(sig: DispatcherSignature) -> Tuple[str, List[Binding]]:
+def unwrap_tensor_args(sig: DispatcherSignature, *, include_sync: bool) -> Tuple[str, List[Binding]]:
```
How should I know to call this with `include_sync=True` versus False?

ezyang replied:
So it looks like the distinction here at the call site is "is it a view or not". I'd suggest naming the parameter after *that*, makes it very easy to understand how you should call it, and then operational consequences of syncing or not are localized to this function.

ezyang commented:
It would be helpful to have a worked example of how two levels of functionalization interact, before and after this change.

ezyang commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -322,6 +322,57 @@ void sync(const c10::List<c10::optional<Tensor>> t_list) {
+    TORCH_INTERNAL_ASSERT(
+         curr_functional == any_functional,
+        "Functionalization encountered a list of tensors where some are functional",
+        "and some are not, which is unsupported.");
```
Unsupported here just means "we didn't bother implement it" as opposed to "fundamentally this is impossible to support" right?

bdhirsh replied:
yep that's right (I'll make the comment a bit clearer).

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -129,11 +136,11 @@ def emit_view_functionalization_body(
 
-    unwrap_tensor_args_str, unwrapped_args_ctx = unwrap_tensor_args(dispatcher_sig)
-    view_redispatch_args = [keyset] + [e.expr for e in translate(unwrapped_args_ctx, call_sig.arguments(), method=False)]
+    # For view ops, we're not using the data from the input tensors so we don't need to sync inputs.
```
oh, commented here

ezyang replied:
It took me a few read throughs to understand this comment. What you appear to be saying is that for views, we don't actually do any compute/copy; we just create a new view object that still points back to the base. So we can continue to defer syncing, since an access to the view object itself will cause a sync. (Importantly, no translation from view to view_copy occurs yet).

bdhirsh replied:
yep, phrased a bit poorly :). I'll make it clearer

bdhirsh replied:
Although technically I don't think we'll need to sync when we convert to `{view}_copy` operators either - we could just copy the stale data, and also copy any of the pending updates from the input `FunctionalTensorWrapper` onto the output. That way when we eventually sync the (stale) output at a later point, all of the previous updates are included too (this is what `pytorch/xla` does today - the optimization where if you create a view and never use the base again, then any mutations on the view don't actually need to get run on the base).

ezyang commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -175,7 +182,7 @@ def emit_view_functionalization_body(
         {meta_conversion_str}
         reference_tensor_output = at::_ops::{api_name}::call({', '.join(meta_call_args)});
-        tmp_output = at::_ops::{api_name}::redispatch({', '.join(view_redispatch_args)});
+        tmp_output = at::_ops::{api_name}::call({', '.join(view_redispatch_args)});
```
This is subtle, probably worth a comment replicating what you said in your PR description

ezyang reviewed (approved):
This looks good, I have some higher level concerns which I'll write about shortly. You should defo let @albanD take a look.

albanD commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -322,6 +322,57 @@ void sync(const c10::List<c10::optional<Tensor>> t_list) {
   }
 }
 
+bool isFunctionalTensor(const at::Tensor& tensor) {
```
TORCH_API missing here I think as it is in the header

zou3519 replied:
meta question: Does TORCH_API need to be on the function definitions? I thought it was only necessary for function declarations

bdhirsh replied:
(yep I think you're right and I added it unnecessarily in a few places - you should need it in the declarations in the header files, but not in the actual cpp files)

albanD commented on aten/src/ATen/FunctionalTensorWrapper.cpp:
```diff
@@ -322,6 +322,57 @@ void sync(const c10::List<c10::optional<Tensor>> t_list) {
+  }
+}
+
+TORCH_API bool isFunctionalTensor(const c10::List<Tensor>& t_list) {
```
The 3 functions on list are exactly the same. Light templating and auto should allow you to merge them no?

bdhirsh replied:
I made a modest attempt at this, but I think it's a little annoying to do - I can't use a generic template for it, since otherwise stuff like `c10::impl::ListElementReference<c10::optional<at::Tensor>>` would incorrectly try to match against it (before, we would implicitly convert to `c10::optional<Tensor>` and call that version).

I think I could fix that using `std::enable_if` to specialize the template for list-like types, but I kind of felt like that's more confusing to look at than just stamping out 3 copies of the 5 line function.

albanD commented on aten/src/ATen/FunctionalizeFallbackKernel.cpp:
```diff
@@ -16,19 +16,23 @@ namespace {
       }
     }
     {
       at::AutoDispatchSkipFunctionalize guard;
```
Is the guard needed here?
Outside of torch_dispatch. Is there any case where a call from a backend might trigger functionalization again?

bdhirsh replied:
yep! (maybe you can get the same behavior by removing the key from the include set here instead, although I just went with adding it to the exclude set).

albanD reviewed (approved):
sgtm as well!

Note that you're missing the run_test.py update to actually run this in CI ;)

ezyang commented:
Broader issues raised at https://github.com/pytorch/pytorch/issues/73513

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```





[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35016024).

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```



Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

[ghstack-poisoned]

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```



Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

[ghstack-poisoned]

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```



Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

[ghstack-poisoned]

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```



Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

[ghstack-poisoned]

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```



Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

[ghstack-poisoned]

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```



Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

[ghstack-poisoned]

bdhirsh commented:
Now that https://github.com/pytorch/pytorch/pull/74577 has landed, I modified this PR a bit and added some tests:

(1) I added some support for mixed inputs: you can pass a functional tensor and a non-functional tensor in, and the output will either be promoted to a functional tensor, or we'll error out if we can't make it work (e.g. `normal_tensor.add_(functional_tensor)` will fail).
(2) Added tests for mixed inputs in `test_functionalization.py`
(3) Added a test for "nested functionalization" in `test_functionalization.py`, similar to what Ed tried to do in https://github.com/pytorch/pytorch/issues/73537

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```



Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

[ghstack-poisoned]

bdhirsh committed:
Update on "make functionalization work better with subclasses"

This PR does two things:

(1) Add `test_functionalization.py` back to CI - at some point it inadvertently got removed, and started failing locally.

The PythonTLSSnapshot PR ended up breaking functionalization (https://github.com/pytorch/pytorch/pull/72623). @albanD helpfully pointed out the fix, and pointed out a useful way to determine when you should use `redispatch/call` to enter the dispatcher. When you're writing a kernel in the dispatcher, you should only `redispatch` if you're redispatching into the *same* operator, with the *same* arguments.

With functionalization, we redispatch into the same operator but with *different* arguments (we unwrap the input tensors, which are `FunctionalTensorWrapper` objects).

The bug: The functionalization tests today are implemented by creating a `LoggingTensor`, converting *that* into a `FunctionalTensorWrapper`, adding the functionalization key to TLS, and then calling a python function. This allows functionalization to run *before* we hit the LoggingTensor backend, so we log the functionalized ops.

With `PythonTLSSnapshot`, we end up (incorrectly) trying to run functionalization twice: we run it before we hit the LoggingTensor impl, but when LoggingTensor redispatches, `PythonTLSSnapshot` adds functionalization back into TLS and we hit it again. This breaks because we've already unwrapped the `FunctionalTensorWrapper` objects, and functionalization kernels aren't currently re-entrant (they raise an assertion error if you're passed a normal `TensorImpl` input).

(2) @albanD also kind of convinced me that the functionalization kernels *should* be re-entrant. I updated the codegen to allow normal tensors to pass through the functionalization kernels, and effectively no-op on tensors that are not "functional".

Admittedly, I'm pretty sure that I didn't need to do that to fix these tests (just switching the `::redispatch`'s to `::call`'s should be enough).

My mental model of how `PythonTLSSnapShot` is meant to work with wrapper tensor subclasses is that functionality keys get re-run at every layer of unwrapping. So in general, you'd expect to see a call-stack like this:
```
myOp():
- [DispatchKey::PythonTLSSnapshot kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey1 kernel] called on outer wrapper tensor
- [DispatchKey::FunctionalityKey2 kernel] called on outer wrapper tensor
- [DispatchKey::Python kernel] called on outer wrapper tensor
* we enter __torch_dispatch__, unwrap the outer tensor and call an op on the inner tensor *
- [DispatchKey::FunctionalityKey1 kernel] called on inner tensor
- [DispatchKey::FunctionalityKey2 kernel] called on inner tensor
- [DispatchKey::CPU kernel] called on inner tensor
```

In that model, it seems reasonable to me that functionalization should be able to handle being called multiple times without breaking.

One potential downside is that we would no longer be able to error on code like this:

```
global_a = torch.ones(...)

def foo():
    global_a.add_(1)

# this won't remove the inplace call, even though we're "functionalizing"
functionalize(foo)()
```



Differential Revision: [D35016024](https://our.internmc.facebook.com/intern/diff/D35016024)

[ghstack-poisoned]

ysiraichi commented on tools/codegen/gen_functionalization_type.py:
```diff
@@ -40,15 +39,26 @@ def is_tensor_like(a: Union[Argument, TensorOptionsArguments, SelfArgument]) ->
+            # a non-view operator.
+            maybe_sync_input = '' if is_view_op else f'at::functionalization::impl::sync({arg.name});'
+            unwrapped_tensor_args.append(f"""
+      {arg.nctype.remove_const_ref().cpp_type()} {unwrapped_name};
+      if (at::functionalization::impl::isFunctionalTensor({arg.name})) {{
+        {maybe_sync_input}
+        {unwrapped_name} = at::functionalization::impl::from_functional_tensor({arg.name});
+      }} else {{
+        {unwrapped_name} = {arg.name};
+      }}""")
```
@ezyang @bdhirsh 
I think I'm having problems related to this change on #73350.

1. Whenever the argument is a `TensorList`, `{arg.nctype.remove_const_ref().cpp_type()}` turns out to be a `TensorList` (not an owning type), too. This will likely lead to segmentation faults.
2. It only compiles because [`from_functional_tensor`](https://github.com/pytorch/pytorch/blob/6642e88ad23bfcf1349f2bce06f3490e49f195bf/aten/src/ATen/FunctionalTensorWrapper.cpp#L298-L304) also returns a `TensorList`. That's a bug, though, since the returned `TensorList` is, in fact, a local `std::vector`.

bdhirsh replied:
Oof, thanks for catching that (and sorry for making you find that bug). Your right, I think that the output of `from_functional_tensor` should always be an owning type (so a vector, in the case where the input was a `TensorList`).

I'm in the middle of making some changes to functionalization, so I can add that fix too and post here when it's ready.

bdhirsh replied:
(fix here: https://github.com/pytorch/pytorch/pull/76125)
